### DISCUSSION TITLE
RPC ClientResponse not picking up response type from zValidator hook

### DISCUSSION THREAD

**Discussion by kosiumeigbo** [honojs/hono#2572] [Q&A]
Hi All,

I am building a project using Hono and NextJS. The version of Hono I am currently using is 4.2.4.

I am using the RPC feature in Hono. I have grouped routes in a file. I am doing validation using the zValidator middleware. The image below is one of the routes/endpoints I have created:

<img width="936" alt="Screenshot 2024-04-29 at 19 06 13" src="https://github.com/honojs/hono/assets/52038224/379f0bdc-5e45-4d21-8f0c-0eb46dbaa5a8">

I have also created a type called 'JSONResponse' to use for all my response objects. If the operation is successful, it returns an object with success true and the data. If the operation is not successful, it returns an object with success false and error property that will be the error message. It helps to keep my response standardised. This is used in the endpoint callback function above. The image below is the type definition for 'JSONResponse':

<img width="460" alt="Screenshot 2024-04-29 at 19 18 19" src="https://github.com/honojs/hono/assets/52038224/d373d024-eb76-4b70-a2b8-d769f440fedc">

I have now imported the hc client to my NextJS server component to use the RPC route. However, I have found that the returned response from my validator hook isn't registered in the Client Response object. It only registers the returned response in the endpoint callback function, which has success as true. It does not register the response in the validation hook. This image is shown below:

<img width="612" alt="Screenshot 2024-04-29 at 19 48 37" src="https://github.com/honojs/hono/assets/52038224/9a4262a4-f2b1-44f5-bc97-fe7cf068b94e">

It should also register the response sent from the zValidator hook so that it would look like something like this:

```js
const res: ClientResponse<{
        success: false;
        error: string
      } | {
        success: true;
        data: ...
      }>
```
Therefore showing a union type between the 'success: true' object and the 'success: false' object. However, this isn't what is done. Am I doing something wrong? Thanks for the help!

**Comment by ambergristle** [comment]
hi @kosiumeigbo! i expect you've already found a solution, but for anyone else running into the same issue:

you're not doing anything wrong; what you want isn't currently supported:

Hono RPC typing only includes `c.json` and `c.text` responses returned from **handlers**. there are a couple open issues addressing this, but it doesn't look like there are concrete plans to resolve:
- https://github.com/honojs/hono/issues/2719
- https://github.com/honojs/hono/issues/3170

**Reply by kosiumeigbo** [reply]
thank you @ambergristle. Yeah, I decided to ditch the RPC client and went full backend by deploying my Hono app on Cloudflare Workers. It seems to be working well so far using this way!

**Reply by ambergristle** [reply]
that makes sense. always nice when you can simplify code!

for those who want/need the RPC client, one option might be to manually type handlers (though imo, at that point why bother): https://hono.dev/docs/guides/rpc#ide-performance

there's also nothing wrong with handling error responses on your own:

```typescript
const response = client.api.books.$get();
if (!response.ok || response.status > 299) {
  // if you have an internal error class, you could instance check here
  // or do a switch/case on the status, etc.
}
```
at the end of the day, error responses are inherently unpredictable. so a client-side implementation that can handle any known/unknown error it might get back is probably preferable to one that only expects a single/few types of error responses

**Reply by samyarkd** [reply]
thanks i was doing `c.notFound` and it messed with the rpc typesafety
